providers/adfs.go

```diff
@@ -0,0 +1,132 @@
+package providers
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+)
+
+// ADFSProvider represents an ADFS based Identity Provider
+type ADFSProvider struct {
+	*OIDCProvider
+}
+
+var _ Provider = (*OIDCProvider)(nil)
+
+const (
+	adfsProviderName = "Adfs"
+	adfsDefaultScope = "openid"
+)
+
+// NewADFSProvider initiates a new AdfsProvider
+func NewADFSProvider(p *OIDCProvider) *ADFSProvider {
+
+	p.ProviderData.setProviderDefaults(providerDefaults{
+		name:  adfsProviderName,
+		scope: adfsDefaultScope,
+	})
+	return &ADFSProvider{OIDCProvider: p}
+}
+
+// GetLoginURL overrides GetLoginURL to add Adfs parameters
+func (p *ADFSProvider) GetLoginURL(redirectURI, state string) string {
+	extraParams := url.Values{}
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
```

Comment:
Are you able to customize the resource & scope that get passed in to be the custom `scope=resource/openid email profile...` format ADFS supports?

That was you can use the existing OIDC implementations of GetLoginURL & Redeem and potentially only need to override EnrichSession with custom claims assignment.

Reply:
Otherwise if you need to make custom versions of every function to splice in `ProtectedResource` (because the golang `oauth2.Config` doesn't support that param) -- you'll also need to do it for the Refresh flow of the OAuth token endpoint in the `RefreshSessionIfNeeded` method.

Reply:
When I tested adding the resource url in the scope it worked fine, so I guess you're right. It's just that I found weird to have the resource param in the config options and not being able to use it for the ADFS provider. Maybe is just enough to add to the docs that for ADFS is enough to pass the resource in the scope.

Reply:
I think you can do special logic in the `NewADFSProvider` method so users can be blissfully unaware of the implementation details. Still take separate `Scope` and `ProtectedResource` arguments -- but if you see a `ProtectedResource` modify the final Scope that this provider gets to be the hybrid structure.